[SYCL][Doc][Graph] Design doc tidy (#361)
reble/llvm: sycl/doc/design/CommandGraph.md

This document describes the implementation design of the
[SYCL Graph Extension](../extensions/experimental/sycl_ext_oneapi_graph.asciidoc).

## Resources

* A recording of a presentation on the extension can be found
  [on Youtube](https://www.youtube.com/watch?v=aOTAmyr04rM).
* A blog post introducing the extension can be found

>   [here](https://codeplay.com/portal/blogs/2024/01/22/sycl-graphs).

Reviewer: ```suggestion
  [on Codeplay.com](https://codeplay.com/portal/blogs/2024/01/22/sycl-graphs).
```
Optional: Might be clearer where this link leads.

![Graph partition illustration step 11b.](images/SYCL-Graph-partitions_step12.jpg)

### Multiple Roots Execution Flow

The following diagram shows the partitions of a graph with two roots and a
host-task in each branch.

![Multiple roots graph partition illustration.](images/SYCL-Graph-multiple_roots_partitions.jpg)

When executing this graph, the partitions were enqueued one after the other,
with each partition waiting for the previous one to complete (see top of the

> following diagram). However, for multi-root graph, this behavior adds

Reviewer 2: ```suggestion
following diagram). However, for a multi-root graph, this behavior adds
```

> unnecessary dependency between partitions, slowing down the execution of the

Reviewer 2: ```suggestion
unnecessary dependencies between partitions, slowing down the execution of the
```

These extra commands operate on L0 event synchronisation primitives, used by
the command-list to interact with the external UR wait-list and UR return event
required for the enqueue interface. Unlike the graph workload (i.e. commands
needed to perform the graph workload) the external UR wait-list and UR return
event are submission dependent, which mean they can change from one submission
to the next.

For performance concerns, the command-list that will execute the graph workload
is made only once (during the command-buffer finalization stage). This allows
the adapter to save time when submitting the command-buffer, by executing only
this command-list (i.e. without enqueuing any commands of the graph workload).

#### Prefix

The prefix's commands aim to:
1. Handle the list of events to wait on, which is passed by the runtime
when the UR command-buffer enqueue function is called. As mentioned above, this
list of events changes from one submission to the next. Consequently, managing
this mutable dependency in the graph-workload command-list implies rebuilding
the command-list for each submission (note that this can change with mutable

> command-list). To avoid the signifiant time penalty of rebuilding this

Reviewer: ```suggestion
command-list). To avoid the significant time penalty of rebuilding this
```

potentially large command-list each time, we prefer to add an extra command
handling the wait list into another command-list (*wait command-list*). This
command-list consists of a single L0 command: a barrier that waits for
dependencies passed by the wait-list and signals a signal called *WaitEvent*
when the barrier is complete. This *WaitEvent* is defined in the
`ur_exp_command_buffer_handle_t` class. In the front of the graph workload
command list, an extra barrier command waiting for this event is added (when
the command-buffer is created). This ensures that the graph workload does not
start running before the dependencies to be completed. The *WaitEvent* event is
reset in the suffix.

2. Reset events associated with the command-buffer except the *WaitEvent* event.
Indeed, L0 events needs to be explicitly reset by an API call  (L0 command in
our case). Since a command-buffer is expected to be submitted multiple times,
we need to ensure that L0 events associated with graph commands have not been

> signalled by a previous execution. These events are therefore reset to the

Reviewer: ```suggestion
signaled by a previous execution. These events are therefore reset to the
```

We use American English in the documentation, right?

Author: Your correct, I completely forgot to check that the reason my editor might be flagging this was that 🤦  I've changed my default language from UK English to US

> non-signalled state before running the graph-workload command-list. Note that

Reviewer: ```suggestion
non-signaled state before running the graph-workload command-list. Note that
```

this reset is performed in the prefix and not in the suffix to avoid additional
synchronization w.r.t profiling data extraction. We use a new command list
(*reset command-list*) for performance concerns.
Indeed:

>    * This allows the *WaitEvent* to be signalled directly on the host if the

Reviewer: ```suggestion
   * This allows the *WaitEvent* to be signaled directly on the host if the
```

   waiting list is empty, thus avoiding the need to submit a command list.
   * Enqueuing a reset L0 command for all events in the command-buffer is time
   consuming, especially for large graphs. However, this task is not needed for
   every submission, but only once, when the command-buffer is fixed, i.e. when
   the command-buffer is finalized. The decorrelation between the reset
   command-list and the wait command-list allow us to create and enqueue the
   reset commands when finalizing the command-buffer, and only create the wait
   command-list at submission.

> This command list consist of a reset command for each of the graph commands

Reviewer 2: ```suggestion
This command list consists of a reset command for each of the graph commands
```

> This command list is consist of a reset command for each of the graph commands

Reviewer: ```suggestion
This command list consists of a reset command for each of the graph commands
```

   executing a UR command-buffer.

3. Dependencies between multiple submissions must be handled by the runtime.
   Indeed, when a second submission is performed the signal conditions of
   *WaitEvent* are redefined by this second submission. Therefore, this can
   lead to an undefined behavior and potential hangs especially if the
   conditions of the first submissions were not yet satisfied and the event has

>    not yet been signalled.

Reviewer: ```suggestion
   not yet been signaled.
```

The SYCL Graph CUDA backend relies on the
[CUDA Graphs feature](https://docs.nvidia.com/cuda/cuda-c-programming-guide/index.html#cuda-graphs),
which is the CUDA public API for batching series of operations, such as kernel
launches, connected by dependencies.

UR commands (e.g. kernels) are mapped as graph nodes using the
[CUDA Driver API](https://docs.nvidia.com/cuda/cuda-driver-api/group__CUDA__GRAPH.html#group__CUDA__GRAPH).
The CUDA Driver API is preferred over the CUDA Runtime API to implement the
SYCL Graph backend to remain consistent with other UR functions.
Synchronization between commands (UR sync-points) is implemented using graph
dependencies.

Executable CUDA Graphs can be submitted to a CUDA stream in the same way as
regular kernels. The CUDA backend enables enqueuing events to wait for into a

> stream. It also allows signalling the completion of a submission with an event.

Reviewer: ```suggestion
stream. It also allows signaling the completion of a submission with an event.
```

Therefore, submitting a UR command-buffer consists only of submitting to a
stream the executable CUDA Graph that represent this series of operations.

An executable CUDA Graph, which contains all commands and synchronization
information, is saved in the UR command-buffer to allow for efficient graph
resubmission.

### HIP

The HIP backend offers a graph management API very similar to CUDA Graph
feature for batching series of operations. The SYCL Graph HIP backend
implementation is therefore very similar to that of CUDA.

UR commands (e.g. kernels) are mapped as graph nodes using the
[HIP Management API](https://docs.amd.com/projects/HIP/en/docs-5.5.0/doxygen/html/group___graph.html).
Synchronization between commands (UR sync-points) is implemented using graph
dependencies. Executable HIP Graphs can be submitted to a HIP stream in the
same way as regular kernels. The HIP backend enables enqueuing events to wait

> for into a stream. It also allows signalling the completion of a submission

Reviewer: ```suggestion
for into a stream. It also allows signaling the completion of a submission
```
